[TT-16831] docs: add OpenTelemetry metrics pages and restructure metrics section (#1808)
TykTechnologies/tyk-docs: api-management/traces/elasticsearch.mdx

---
title: "Send Tyk Traces to Elasticsearch"
description: "Step-by-step guide to connect Tyk Gateway distributed traces to Elasticsearch using the OpenTelemetry Collector, for OSS, self-managed, and hybrid deployments."
keywords: "Distributed Tracing, Elasticsearch, Elastic APM, OpenTelemetry, OTLP, OTel Collector, Tyk Gateway, Observability"
sidebarTitle: "Elasticsearch"
---

> This quick start explains how to configure Tyk API Gateway (OSS, self-managed or hybrid gateway connected to Tyk Cloud) with the OpenTelemetry Collector to export distributed traces to [Elasticsearch](https://www.elastic.co/observability).

Reviewer: ```suggestion
This guide explains how to set up [Elasticsearch](https://www.elastic.co/observability) to ingest OpenTelemetry traces via the OpenTelemetry Collector (OTel Collector).
```

## Prerequisites

Ensure the following prerequisites are met before proceeding:

> * Tyk Gateway v5.2 or higher

Reviewer: ```suggestion
- Tyk Gateway v5.2 or higher
```

> * OpenTelemetry Collector deployed locally

Reviewer: ```suggestion
- OpenTelemetry Collector deployed locally
```

> * Elasticsearch deployed locally or an account on Elastic Cloud with Elastic APM

Reviewer: ```suggestion
- Elasticsearch deployed locally or an account on Elastic Cloud with Elastic APM
```

Elastic Observability natively supports OpenTelemetry and its OpenTelemetry protocol (OTLP) to ingest traces, metrics, and logs. 

<img src="/img/distributed-tracing/opentelemetry/elastic-otel.png" alt="OpenTelemetry support in Elasticsearch" />
Credit: Elasticsearch, [OpenTelemetry on Elastic](https://www.elastic.co/blog/opentelemetry-observability)

## Steps for Configuration 

> ### Step 1. Configure Tyk API Gateway

Reviewer: ```suggestion
### Step 1. Configure Tyk Gateway
```

>     To enable OpenTelemetry in Tyk API Gateway, follow these steps:

Reviewer: ```suggestion
```

>     For Tyk Helm Charts:

Reviewer: ```suggestion
    To enable OpenTelemetry when using Tyk Helm Charts add the following configuration to the Tyk Gateway section:
```

>     * Add the following configuration to the Tyk Gateway section:

Reviewer: ```suggestion
```

    ```yaml
    tyk-gateway:
      gateway:
        opentelemetry:
          enabled: true
          endpoint: {{Add your endpoint here}}
          exporter: grpc
    ```

>     For Docker Compose:

Reviewer: ```suggestion
    To enable OpenTelemetry when using Docker Compose add the following environment variables to your `docker-compose.yml` file for Tyk Gateway:
```

>     * In your docker-compose.yml file for Tyk Gateway, add the following environment variables:

Reviewer: ```suggestion
```

    ```yaml
    environment:
      - TYK_GW_OPENTELEMETRY_ENABLED=true
      - TYK_GW_OPENTELEMETRY_EXPORTER=grpc
      - TYK_GW_OPENTELEMETRY_ENDPOINT={{Add your endpoint here}}
    ```

>     Make sure to replace `<Add your endpoint here>` with the appropriate endpoint from your OpenTelemetry collector.

Reviewer: ```suggestion
<Note>
    For both deployment types, make sure to replace `<Add your endpoint here>` with the appropriate endpoint from your OpenTelemetry collector.
</Note>
```

>     After enabling OpenTelemetry at the Gateway level, you can activate [detailed tracing](/api-management/traces#opentelemetry) for specific APIs by editing their respective API definitions. Set the `detailed_tracing` option to either true or false. By default, this setting is false.

Reviewer: ```suggestion
    After enabling OpenTelemetry for the Gateway, you can activate [detailed tracing](/api-management/traces#opentelemetry) for specific APIs in their respective API definitions. Detailed tracing is not enabled by default so you will need to set the `detailed_tracing` option to `true` to collect detailed traces.
```

### Step 2. Configure the OpenTelemetry Collector to Export to Elasticsearch

    To configure the OTel Collector with Elasticsearch Cloud, follow these steps:

>     * Sign up for an [Elastic account](https://www.elastic.co/) if you haven't already

Reviewer: ```suggestion
    1. Sign up for an [Elastic account](https://www.elastic.co/) if you haven't already
```

>     * Once logged in to your Elastic account, select "Observability" and click on the option "Monitor my application performance"

Reviewer: ```suggestion
    2. Once logged in to your Elastic account, select **Observability** and click on the option **Monitor my application performance**
```

    <img src="/img/distributed-tracing/opentelemetry/elastic-configure-otel-01.png" alt="Configure Elasticsearch" />

>     * Scroll down to the APM Agents section and click on the OpenTelemetry tab

Reviewer: ```suggestion
    3. Scroll down to the **APM Agents** section and click on the **OpenTelemetry** tab
```

    <img src="/img/distributed-tracing/opentelemetry/elastic-configure-otel-02.png" alt="Configure Elasticsearch" />

>     * Search for the section "Configure OpenTelemetry in your application". You will need to copy the value of "OTEL_EXPORTER_OTLP_ENDPOINT" and "OTEL_EXPORTER_OTLP_HEADERS" in your OpenTelemetry Collector configuration file.

Reviewer: ```suggestion
    4. Search for the section **Configure OpenTelemetry in your application"** You will need to copy the value of `OTEL_EXPORTER_OTLP_ENDPOINT` and `OTEL_EXPORTER_OTLP_HEADERS` in your OpenTelemetry Collector configuration file.
```

    <img src="/img/distributed-tracing/opentelemetry/elastic-configure-otel-03.png" alt="Configure Elasticsearch" />

>     * Update your OpenTelemetry Collector configuration, here's a simple example:

Reviewer: ```suggestion
    5. Update your OpenTelemetry Collector configuration, here's a simple example:
```

    ```yaml expandable
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: 0.0.0.0:4317 # OpenTelemetry receiver endpoint
    processors:
      batch:
    exporters:
    otlp/elastic:
        endpoint: "ELASTIC_APM_SERVER_ENDPOINT_GOES_HERE" #exclude scheme, e.g. HTTPS:// or HTTP://
        headers:
          # Elastic APM Server secret token
          Authorization: "Bearer ELASTIC_APM_SECRET_TOKEN_GOES_HERE"
    service:
      pipelines:
        traces:
          receivers: [otlp]
          exporters: [otlp/elastic]
    ```

>     If are running Elasticsearch locally, you will need to use your APM Server endpoint (elastic-apm-server:8200) and set-up [a secret token authorization in ElasticSearch](https://www.elastic.co/guide/en/observability/current/secret-token.html).

Reviewer: ```suggestion
    If are running Elasticsearch locally, you will need to use your APM Server endpoint (elastic-apm-server:8200) and set up [a secret token authorization in ElasticSearch](https://www.elastic.co/guide/en/observability/current/secret-token.html).
```

    You can refer to the [example configuration provided by Elastic](https://www.elastic.co/guide/en/observability/current/open-telemetry-direct.html#connect-open-telemetry-collector) for more guidance on the OpenTelemetry Collector configuration.

### Step 3. Explore OpenTelemetry Traces in Elasticsearch

>     * In Elasticsearch Cloud:

Reviewer: ```suggestion
    In Elasticsearch Cloud:
```

>     * Go to "Home" and select "Observability."

Reviewer: ```suggestion
    1. Go to **Home** and select **Observability**.
```

    <img src="/img/distributed-tracing/opentelemetry/elastic-configure-otel-04.png" alt="Configure Elasticsearch" />

>     * On the right menu, click on "APM / Services."

Reviewer: ```suggestion
    2. On the right menu, click on **APM / Services**.
```

>     * Click on "tyk-gateway."

Reviewer: ```suggestion
    3. Click on **tyk-gateway**.
```

>     You will see a dashboard automatically generated based on the distributed traces sent by Tyk API Gateway to Elasticsearch.

Reviewer: ```suggestion
    4. You will see a dashboard automatically generated based on the distributed traces sent by Tyk Gateway to Elasticsearch.
```
